### PR TITLE
Video projection selection via URL parameter

### PR DESCRIPTION
**Pull request by MortimerGoro**
There is currently no API in GV that allows to enter fullscreen automatically.

At least this PR enters the video projection when users fullscreens it.

Example URL:
https://www.youtube.com/watch?v=sPyAQQklc1s&mozVideoProjection=360

### CODE REVIEW

File: app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java [hunk 1]
```diff
@@ -116,4 +117,37 @@ private void handleClick(@VideoProjectionFlags int aVideoProjection) {
...
+
+        if (projection.equals("360")) {
+            return VIDEO_PROJECTION_360;
+        } else if (projection.equals("360s")) {
+            return VIDEO_PROJECTION_360_STEREO;
+        } else if (projection.equals("180")) {
+            return VIDEO_PROJECTION_180;
+        } else if (projection.equals("180lr")) {
+            return VIDEO_PROJECTION_180_STEREO_LEFT_RIGTH;
+        } else if (projection.equals("180tp")) {
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java]: typo: `p` should be `b`

File: app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java [hunk 2]
```diff
@@ -116,4 +117,37 @@ private void handleClick(@VideoProjectionFlags int aVideoProjection) {
...
+        if (aURL == null) {
+            return -1;
+        }
+
+        Uri uri = Uri.parse(aURL);
+        if (uri == null) {
+            return -1;
+        }
+
+        String projection = uri.getQueryParameter("mozVideoProjection");
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java]: is there a way to allow the key name to be case insensitive? without iterating through all of the query-strings, perhaps also check `mozvideoprojection`?

File: app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java [hunk 3]
```diff
@@ -116,4 +117,37 @@ private void handleClick(@VideoProjectionFlags int aVideoProjection) {
...
+        projection = projection.toLowerCase();
+
+        if (projection.equals("360")) {
+            return VIDEO_PROJECTION_360;
+        } else if (projection.equals("360s")) {
+            return VIDEO_PROJECTION_360_STEREO;
+        } else if (projection.equals("180")) {
+            return VIDEO_PROJECTION_180;
+        } else if (projection.equals("180lr")) {
+            return VIDEO_PROJECTION_180_STEREO_LEFT_RIGTH;
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java]: typo: `RIGTH` should be `RIGHT` here and elsewhere in the file

File: app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java [hunk 4]
```diff
@@ -116,4 +117,37 @@ private void handleClick(@VideoProjectionFlags int aVideoProjection) {
...
+        } else if (projection.equals("180tp")) {
+            return VIDEO_PROJECTION_180_STEREO_TOP_BOTTOM;
+        } else if (projection.equals("3d")) {
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java]: what about 3D top bottom?

File: app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/NavigationBarWidget.java [hunk 5]
```diff
@@ -699,7 +699,10 @@ public void onFullScreen(GeckoSession session, boolean aFullScreen) {
             if (mIsResizing) {
                 exitResizeMode(false);
             }
-
+            int projection = VideoProjectionMenuWidget.getAutomaticProjection(SessionStore.get().getUriFromSession(session));
+            if (projection >= 0) {
+                enterVRVideo(projection);
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/NavigationBarWidget.java]: I like this - as opposed to automatically presenting when the page loads. smart thinking :+1:

**Comment by cvan** [comment]
@caseyyee: when you have a moment tomorrow, could you check this with your local build of FxR in the headset?

here's a flow to test of relevant videos, in order:

1. https://www.youtube.com/watch?v=sPyAQQklc1s&mozVideoProjection=360
2. https://www.youtube.com/watch?v=4dwjS_eI-lQ
3. https://www.youtube.com/watch?v=sPyAQQklc1s

thanks!

**Comment by cvan** [comment]
(testing with this branch vs. master) when I traverse away, I notice the Projection Mode UI list persists when I press the Back button on the controller

**Comment by cvan** [comment]
I tried to fix the typos but instead the functionality no longer seems to work (I did a fresh Clean, Build, and Run).

any ideas?

```diff
diff --git a/app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java b/app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java
index 67218f6..1ba0f87 100644
--- a/app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java
+++ b/app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java
@@ -11,16 +11,19 @@ import java.util.ArrayList;
 
 public class VideoProjectionMenuWidget extends MenuWidget {
 
-    @IntDef(value = { VIDEO_PROJECTION_3D_SIDE_BY_SIDE, VIDEO_PROJECTION_360,
-                      VIDEO_PROJECTION_360_STEREO, VIDEO_PROJECTION_180,
-                      VIDEO_PROJECTION_180_STEREO_LEFT_RIGTH, VIDEO_PROJECTION_180_STEREO_TOP_BOTTOM })
+    @IntDef(value = { VIDEO_PROJECTION_3D_SIDE_BY_SIDE,
+      VIDEO_PROJECTION_360,
+      VIDEO_PROJECTION_360_STEREO,
+      VIDEO_PROJECTION_180,
+      VIDEO_PROJECTION_180_STEREO_LEFT_RIGHT,
+      VIDEO_PROJECTION_180_STEREO_TOP_BOTTOM })
     public @interface VideoProjectionFlags {}
 
     public static final int VIDEO_PROJECTION_3D_SIDE_BY_SIDE = 0;
     public static final int VIDEO_PROJECTION_360 = 1;
     public static final int VIDEO_PROJECTION_360_STEREO = 2;
     public static final int VIDEO_PROJECTION_180 = 3;
-    public static final int VIDEO_PROJECTION_180_STEREO_LEFT_RIGTH = 4;
+    public static final int VIDEO_PROJECTION_180_STEREO_LEFT_RIGHT = 4;
     public static final int VIDEO_PROJECTION_180_STEREO_TOP_BOTTOM = 5;
 
     public interface Delegate {
@@ -90,7 +93,7 @@ public class VideoProjectionMenuWidget extends MenuWidget {
         mItems.add(new MenuItem(R.string.video_mode_180_left_right, R.drawable.ic_icon_videoplayback_180_stereo_leftright, new Runnable() {
             @Override
             public void run() {
-                handleClick(VIDEO_PROJECTION_180_STEREO_LEFT_RIGTH);
+                handleClick(VIDEO_PROJECTION_180_STEREO_LEFT_RIGHT);
             }
         }));
 
@@ -141,8 +144,8 @@ public class VideoProjectionMenuWidget extends MenuWidget {
         } else if (projection.equals("180")) {
             return VIDEO_PROJECTION_180;
         } else if (projection.equals("180lr")) {
-            return VIDEO_PROJECTION_180_STEREO_LEFT_RIGTH;
-        } else if (projection.equals("180tp")) {
+            return VIDEO_PROJECTION_180_STEREO_LEFT_RIGHT;
+        } else if (projection.equals("180tb")) {
             return VIDEO_PROJECTION_180_STEREO_TOP_BOTTOM;
         } else if (projection.equals("3d")) {
             return VIDEO_PROJECTION_3D_SIDE_BY_SIDE;
```

File: app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java (hunk 4 above)
Comment (caseyyee) [reply app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/VideoProjectionMenuWidget.java]: Usually referred to as "over-under". 
Also Left/Right is usually referred to "side-by-side" or _SBS_

We shouldn't refer to video as 3D.  It's stereo 360 or stereo 180, both being typically SBS (mostly assumed).

**Comment by MortimerGoro** [comment]
Autoenter is a bit broken after the UA change, probably due to the Youtube layout problems. I added an option to separately decide if we want to autoselect the projection when clicking the icon or enter automatically when clicking fullscreen.

Automatically selects the projection on VR icon click:
https://www.youtube.com/watch?v=sPyAQQklc1s&mozVideoProjection=360

Automatically selects the projection and enters VR video when fullscreened:
https://www.youtube.com/watch?v=sPyAQQklc1s&mozVideoProjection=360_auto

**Pull request merged**